File: solutions.digamma.damas.jcr/src/main/java/solutions/digamma/damas/jcr/content/JcrFolder.java
package solutions.digamma.damas.jcr.content;

import solutions.digamma.damas.common.WorkspaceException;
import solutions.digamma.damas.common.InternalStateException;
import solutions.digamma.damas.content.DetailedFolder;
import solutions.digamma.damas.content.Folder;
import solutions.digamma.damas.inspection.NotNull;
import solutions.digamma.damas.jcr.names.TypeNamespace;
import solutions.digamma.damas.jcr.common.Exceptions;

import javax.jcr.Node;
import javax.jcr.NodeIterator;
import javax.jcr.RepositoryException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * JCR-based folder implementation.
 *
 * @author Ahmad Shahwan
 */
public class JcrFolder extends JcrFile implements Folder {

    /**
     * Content depth used to indicate to which depth is content returned with
     * method {@code getContent()}. If negative, content will be returned down
     * to leaves. If zero, no content is returned at all. Zero is the default
     * value.
     */
    private int contentDepth = 0;

    /**
     * Folder content. If null, the content is not visited yet.
     */
    private Content content;

    /**
     * Construct folder with a JCR node.
     *
     * @param node
     */
    JcrFolder(@NotNull Node node) throws WorkspaceException {
        super(node);
    }

    @Override
    protected void checkCompatibility() throws InternalStateException {
        super.checkCompatibility();
        this.checkTypeCompatibility(TypeNamespace.FOLDER);
    }

    @Override
    public @NotNull DetailedFolder expand() throws WorkspaceException {
        return new JcrDetailedFolder(this.node);
    }

    @Override
    public void expandContent(int depth) {
        if (depth != this.contentDepth) {
            this.content = null;
            this.contentDepth = depth;
        }
    }

    @Override
    public void expandContent() {
        if (this.contentDepth >= 0) {
            this.content = null;
            this.contentDepth = -1;
        }
    }

    @Override
    public Content getContent() throws WorkspaceException {
        if (this.contentDepth == 0) {
            return this.content = null;
        }
        if (this.content != null) {
            return this.content;
        }
        List<JcrDocument> documents = new ArrayList<>();
        List<JcrFolder> folders = new ArrayList<>();
        try {
            NodeIterator iterator = this.node.getNodes();
            while (iterator.hasNext()) {
                Node node = iterator.nextNode();
                if (node.isNodeType(TypeNamespace.DOCUMENT)) {
                    documents.add(new JcrDocument(node));
                } else if (node.isNodeType(TypeNamespace.FOLDER)) {
                    JcrFolder folder = new JcrFolder(node);
                    folder.expandContent(this.contentDepth - 1);
                    folders.add(folder);
                }
            }
        } catch (RepositoryException e) {
            throw Exceptions.convert(e);
        }
        return this.content = new Content() {
            @Override
            public @NotNull List<JcrFolder> getFolders() {
                return Collections.unmodifiableList(folders);
            }

            @Override
            public @NotNull List<JcrDocument> getDocuments() {
                return Collections.unmodifiableList(documents);
            }
        };
    }
}

File: solutions.digamma.damas.jcr/src/main/kotlin/solutions/digamma/damas/jcr/content/JcrFolder.kt
package solutions.digamma.damas.jcr.content

import solutions.digamma.damas.common.InternalStateException
import solutions.digamma.damas.common.WorkspaceException
import solutions.digamma.damas.content.Folder
import solutions.digamma.damas.jcr.common.Exceptions
import solutions.digamma.damas.jcr.names.TypeNamespace
import java.util.Collections
import javax.jcr.Node
import javax.jcr.Session

/**
 * JCR-based folder implementation.
 *
 * @constructor Private construct folder with a JCR node.
 * @param node
 *
 * @author Ahmad Shahwan
 */
internal open class JcrFolder
@Throws(WorkspaceException::class)
protected constructor(node: Node) : JcrFile(node), Folder {

    /**
     * Content depth used to indicate to which depth is content returned with
     * method `getContent()`. If negative, content will be returned down
     * to leaves. If zero, no content is returned at all. Zero is the default
     * value.
     */
    private var contentDepth = 0

    /**
     * Folder content. If null, the content is not visited yet.
     */
    private var content: Folder.Content? = null

    @Throws(InternalStateException::class)
    override fun checkCompatibility() {
        super.checkCompatibility()
        this.checkTypeCompatibility(TypeNamespace.FOLDER)
    }

    /**
     * Make a copy of the current document under the folder identified by the
     * given parent ID.
     *
     * @param parentId      the copy's parent ID
     * @return              newly created document
     * @exception WorkspaceException
     */
    @Throws(WorkspaceException::class)
    fun duplicate(parentId: String): JcrFolder {
        return JcrFolder.of(session.getNode(copyUnder(parentId)))
    }

    override fun expandContent(depth: Int) {
        if (depth != this.contentDepth) {
            this.content = null
            this.contentDepth = depth
        }
    }

    override fun expandContent() {
        if (this.contentDepth >= 0) {
            this.content = null
            this.contentDepth = -1
        }
    }

    override fun getContent(): Folder.Content? {
        if (this.contentDepth == 0) {
            this.content = null
            return this.content
        }
        if (this.content != null) {
            return this.content
        }
        val documents = ArrayList<JcrDocument>()
        val folders = ArrayList<JcrFolder>()
        Exceptions.uncheck {
            val iterator = this.node.nodes
            while (iterator.hasNext()) {
                val node = iterator.nextNode()
                if (node.isNodeType(TypeNamespace.DOCUMENT)) {
                    documents.add(JcrDocument.of(node))
                } else if (node.isNodeType(TypeNamespace.FOLDER)) {
                    val folder = JcrFolder.of(node)
                    folder.expandContent(this.contentDepth - 1)
                    folders.add(folder)
                }
            }
        }

        this.content = object : Folder.Content {

            override fun getFolders() =
                    Collections.unmodifiableList(folders)

            override fun getDocuments() =
                    Collections.unmodifiableList(documents)
        }
        return this.content
    }

    companion object {

        @Throws(WorkspaceException::class)
        fun of(node: Node) = JcrFolder(node)

        @Throws(WorkspaceException::class)
        fun from(session: Session, parentId: String, name: String) =
                Exceptions.check {
            val parent = session.getNodeByIdentifier(parentId)
            if (parent.hasNode(name)) {
                throw FileExistsException(name)
            }
            val node = parent.addNode(name, TypeNamespace.FOLDER)
            node.addMixin(TypeNamespace.FILE)
            JcrFolder.of(node).apply { initPermissions() }
        }
    }
}
